File: backend/app/services/review.py
from uuid import UUID
from typing import Optional, List
from app.models import ReviewInDB, ReviewCreate, ReviewUpdate, ReviewResponse
from app.repository import ReviewRepository
from app.utils.exceptions import (
    NotFoundException,
    InternalServerException,
    ConfictException,
)


def review_in_db_to_response(review: ReviewInDB) -> ReviewResponse:
    """"""
    return ReviewResponse(**review.model_dump())


class ReviewService:
    def __init__(self, repo: ReviewRepository) -> None:
        self.repo = repo

    async def create_review(self, data: ReviewCreate) -> ReviewResponse:
        """"""
        if await self.repo.exists(data.booking_id):
            raise ConfictException("You have already reviewed this booking")
        review: Optional[ReviewInDB] = await self.repo.create(data.model_dump())
        if review is None:
            raise InternalServerException("Error creating review")
        return review_in_db_to_response(review)

    async def readone_review(self, review_id: UUID) -> ReviewResponse:
        """"""
        review: Optional[ReviewInDB] = await self.repo.readone(review_id)
        if review is None:
            raise NotFoundException("Review not found")
        return review_in_db_to_response(review)

    async def readall_reviews(
        self,
        client_id: Optional[UUID] = None,
        technician_id: Optional[UUID] = None,
        min_rating: Optional[int] = None,
        skip: int = 0,
        limit: int = 100,
    ) -> List[ReviewResponse]:
        """"""
        reviews: List[ReviewInDB] = await self.repo.readall(
            client_id, technician_id, min_rating, skip, limit
        )
        return [review_in_db_to_response(review) for review in reviews]

    async def update_review(self, review_id: UUID, data: ReviewUpdate) -> ReviewResponse:
        """"""
        review: Optional[ReviewInDB] = await self.repo.readone(review_id)
        if review is None:
            raise NotFoundException("Review not found")
        review: Optional[ReviewInDB] = await self.repo.update(
            review_id, data.model_dump()
        )
        if review is None:
            raise InternalServerException("Error updating review")
        return review_in_db_to_response(review)

    async def delete_review(self, review_id: UUID) -> bool:
        """"""
        return await self.repo.delete(review_id)

File: frontend/src/lib/services/review.ts
import axios from "axios";
import { BASE_API_URL } from "@/lib/constants";
import type { ReviewCreate, ReviewUpdate, ReviewResponse } from "@/lib/types/review";

const URL = `${BASE_API_URL}/review`;

export async function createReview(data: ReviewCreate): Promise<ReviewResponse> {
    const response = await axios.post(`${URL}/`, data);
    return response.data;
}

export async function readoneReview(review_id: string): Promise<ReviewResponse> {
    const response = await axios.get(`${URL}/${review_id}`);
    return response.data;
}

export async function readallReviews(
    client_id?: string,
    technician_id?: string,
    min_rating?: number,
    skip: number = 0,
    limit: number = 100
): Promise<ReviewResponse[]> {
    const params: { [key: string]: string | number } = {};

    if (client_id !== undefined) {
        params.client_id = client_id;
    }
    if (technician_id !== undefined) {
        params.technician_id = technician_id;
    }
    if (min_rating !== undefined) {
        params.min_rating = min_rating;
    }
    params.skip = skip;
    params.limit = limit;

    const response = await axios.get(URL, { params });
    return response.data;
}

export async function updateReview(review_id: string, data: ReviewUpdate): Promise<ReviewResponse> {
    const response = await axios.put(`${URL}/${review_id}`, data);
    return response.data;
}

export async function deleteReview(review_id: string): Promise<boolean> {
    const response = await axios.delete(`${URL}/${review_id}`);
    return response.data;
}
